core/modules/qdisp/testQDisp.cc

```diff
@@ -0,0 +1,412 @@
+	boost::shared_ptr<qdisp::MessageStore> ms = boost::make_shared<qdisp::MessageStore>();
+	qdisp::Executive ex(conf, ms);
+	SequentialInt sequence(0);
+	SequentialInt chunkId(1234);
+	int jobs = 0;
+	// test single instance
+	int millisInt = 400;
+	std::string millis(boost::lexical_cast<std::string>(millisInt));
+	double duration = executiveTest(ex, sequence, chunkId, millis, 1);
+	++jobs;
+	LOGF_INFO("duration single %1%" % duration);
+	BOOST_CHECK(ex._empty.get() == true);
+	// test adding 4 jobs
+	duration = executiveTest(ex, sequence, chunkId, millis, 4);
+	jobs += 4;
+	LOGF_INFO("duration four %1%" % duration);
+	BOOST_CHECK(ex._empty.get() == true);
+	// test multiple threads
+	int threadCount = 10;
+	int jobsPerThread = 2;
+	std::vector<WorkerTest> workers(threadCount);
+	std::vector<boost::shared_ptr<boost::thread> > threads(threadCount);
+	// Run half with long sleep value.
+	int j = 0;
+	for(; j<threadCount/2; ++j) {
+		threads[j] = boost::make_shared<boost::thread>(&WorkerTest::operator(), &(workers[j]), boost::ref(ex), boost::ref(sequence),
+                boost::ref(chunkId), millis, jobsPerThread);
+		jobs += jobsPerThread;
+	}
+	// Run the rest with short sleep value
+	qdisp::XrdSsiServiceMock::_go.set(false); // prevent any new threads from completing until after next test.
+	std::string millisShort("10");
+	for(; j<threadCount; ++j) {
+		threads[j] = boost::make_shared<boost::thread>(&WorkerTest::operator(), &(workers[j]), boost::ref(ex), boost::ref(sequence),
+				boost::ref(chunkId), millisShort, jobsPerThread);
+		jobs += jobsPerThread;
+	}
+	boost::thread timeoutT(&timeoutFunc, boost::ref(ex), jobs, millisInt*10);
+
+	// Verify that we can detect jobs have been added but not completed (must make sure jobs have been added).
+	while (qdisp::XrdSsiServiceMock::_count.get() < jobs) {
+		LOGF_INFO("waiting for _count(%1%) == jobs(%2%)" % qdisp::XrdSsiServiceMock::_count.get() % jobs);
+		boost::this_thread::sleep(boost::posix_time::milliseconds(10));
+	}
+	BOOST_CHECK(ex._empty.get() == false);
+
+	qdisp::XrdSsiServiceMock::_go.set(true);
+	for(int j=0; j<threadCount; ++j) {
+		threads[j]->join();
+		LOGF_INFO("Thread %1% duration %2%" % j % workers[j]._retVal);
+	}
+	timeoutT.join();
+	LOGF_INFO("ex._requestCount=%1% jobs=%2% empty=%3%" % ex._requestCount % jobs % ex._empty.get());
+	BOOST_CHECK(ex._empty.get() == true);
+	LOGF_INFO("Executive test end");
+
+}
+
+BOOST_AUTO_TEST_CASE(MessageStore) {
+	LOGF_INFO("MessageStore test start");
+	qdisp::MessageStore ms;
+	BOOST_CHECK(ms.messageCount() == 0);
+	ms.addMessage(123, 456, "test1");
+	std::string str("test2");
+	ms.addMessage(124, -12, str);
+	ms.addMessage(86, -12, "test3");
+	BOOST_CHECK(ms.messageCount() == 3);
+	BOOST_CHECK(ms.messageCount(-12) == 2);
+	qdisp::QueryMessage qm = ms.getMessage(1);
+	BOOST_CHECK(qm.chunkId == 124 && qm.code == -12 && 	str.compare(qm.description) == 0);
+	LOGF_INFO("MessageStore test end");
+}
+
+BOOST_AUTO_TEST_CASE(QueryResource) {
```

Comment:
Wouldn't it be better to mock the requester, so you don't need an executive, a chunkmsgreceiver, a merger, etc.? It's best not to involve an Executive instance if you don't have to.